serialization/serialization.go

```diff
@@ -0,0 +1,283 @@
+type FieldIntVar struct {
+	Value int64
+}
+
+func (f FieldIntVar) String() string {
+	return fmt.Sprintf("%d", f.Value)
+}
+
+// FieldUintVar is using the usigned integer variant of the 'varlen_integer_format'
+// and encodes a value as a byte sequence of 1-9 bytes depending on the value.
+type FieldUintVar struct {
+	Value uint64
+}
+
+func (f FieldUintVar) String() string {
+	return fmt.Sprintf("%d", f.Value)
+}
+
+// FieldString is a 'string_format' field
+type FieldString struct {
+	Value string
+}
+
+func (f FieldString) String() string {
+	return f.Value
+}
+
+func Unmarshal(data []byte, v interface{}) error {
+	switch m := v.(type) {
+	case *Message:
+		m.Version = data[0] >> 1
+		err := Unmarshal(data[1:], &m.Format)
+		if err != nil {
+			return err
+		}
+	case *Format:
+		pos := uint64(0)
+		m.Size = uint64(data[pos] >> 1)
+		pos++
+		m.LastNonIgnorableField = int(data[pos] >> 1)
+		pos++
+
+		for i := 0; i < len(m.Fields); i++ {
+			if int(pos)+1 > len(data) || int(data[pos]>>1) != i {
+				// The field number we got doesn't match what we expect,
+				// so a field was skipped.
+				m.Fields[i].ID = i
+				m.Fields[i].Skipped = true
+				continue
+			}
+			m.Fields[i].ID = int(data[pos] >> 1)
+			pos++
+			var n uint64
+			var err error
+			switch f := m.Fields[i].Type.(type) {
+			case FieldIntFixed:
+				f.Value, n, err = decodeFixed(data, pos, f.Length)
+				if err != nil {
+					return err
+				}
+				m.Fields[i].Type = f
+			case FieldUintVar:
+				var val interface{}
+				val, n, err = decodeVar(data, pos, true)
+				if err != nil {
+					return err
+				}
+				if uintval, ok := val.(uint64); ok {
+					f.Value = uintval
+				} else {
+					return errors.New("unexpected type, expecting uint64")
+				}
+				m.Fields[i].Type = f
+			case FieldIntVar:
+				var val interface{}
+				val, n, err = decodeVar(data, pos, false)
+				if err != nil {
+					return err
+				}
+				if intval, ok := val.(int64); ok {
+					f.Value = intval
+				} else {
+					return errors.New("unexpected type, expecting int64")
+				}
+				m.Fields[i].Type = f
+			case FieldString:
+				f.Value, n, err = decodeString(data, pos)
+				if err != nil {
+					return err
+				}
+				m.Fields[i].Type = f
+			default:
+				return fmt.Errorf("unsupported field type: %T", m.Fields[i].Type)
+			}
+			pos = n
+		}
+
+	default:
+		return fmt.Errorf("unsupported type: %T", v)
+	}
+	return nil
+}
+
+func decodeString(data []byte, pos uint64) (string, uint64, error) {
+	if len(data) < int(pos)+1 {
+		return "", pos, errors.New("string truncated, expected at least one byte")
+	}
+	strLen := int(data[pos] >> 1)
+	pos++
+	if len(data) < int(pos)+strLen {
+		return "", pos, fmt.Errorf("string truncated, expected length: %d", strLen)
+	}
+	return string(data[pos : pos+uint64(strLen)]), pos + uint64(strLen), nil
+}
+
+func decodeFixed(data []byte, pos uint64, intlen int) ([]byte, uint64, error) {
+	var b bytes.Buffer
```

Comment:
```suggestion
	var b bytes.Buffer
	b.Grow(intlen)
```

Reply:
Hm. The fixed input might lead to variable output between `intlen` and `intlen*2`. So `intlen` is often not enough.

Reply:
`b.Grow(intlen * 2)` would be suitable in that case

Ran a test on playground, cap is initialized to 8, then grows to 16